interceptor/reflection.go

```diff
@@ -40,19 +44,47 @@ var (
 	}
 )
 
-// stringMatcher returns 2 values:
-//  1. new name. If there is no change, new name equals to input name
-//  2. whether or not the input name matches the defined rule(s).
-type stringMatcher func(name string) (string, bool)
+type (
+	// Visitor will visits an object's fields recursively. It returns an
+	// implementation-specific bool and error, which typicall indicate if it
+	// matched anything and if it encountered an unrecoverable error.
+	Visitor interface {
+		Visit(any) (bool, error)
+	}
+
+	// visitNamespace uses reflection to recursively visit all fields in the
+	// given object. When it finds namespace string fields, it invokes the match
+	// function.
+	nsVisitor struct {
+		match stringMatcher
+	}
+
+	// saVisitor uses reflection to recursively visit search attribute fields in the given object.
+	// It translates search attribute fields according to per-namespace search attribute mappings.
+	//
+	// This is not concurrent safe. You must create a separate struct each time.
+	saVisitor struct {
+		getNamespaceSAMatcher getSAMatcher
 
-// visitor visits each field in obj matching the matcher.
-// It returns whether anything was matched and any error it encountered.
-type visitor func(obj any, match stringMatcher) (bool, error)
+		// currentNamespaceId is internal-state to remember the namespace id set in some parent
+		// field as the visitor descends recursively into child fields.
+		currentNamespaceId string
```

Comment:
Switch to a Visitor struct so that we can track `currentNamespaceId` as we recursively descend into struct.